### PR TITLE
Persist feedback

### PR DESCRIPTION
**Pull request by shachar-langer**

### CODE REVIEW

File: db/migrations/20201109145014_add_feedbacks_table.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
+const { addUpdateTsTrigger } = require("../util")
+
+exports.up = async (knex) => {
+  await knex.schema.createTable("feedbacks", (table) => {
```
Comment (will-weiss) [inline db/migrations/20201109145014_add_feedbacks_table.js]: Let's just name this `feedback`, which is both the singular and plural

File: db/migrations/20201109145014_add_feedbacks_table.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    table.increments()
+    table
+      .integer("user_id")
+      .notNullable()
+      .references("id")
+      .inTable("users")
+      .onDelete("CASCADE")
+      .index()
+    table.string("status").notNullable()
```
Comment (will-weiss) [inline db/migrations/20201109145014_add_feedbacks_table.js]: We should include a reference to nonnullable reference to `website_id`, whose domain should be included in the incoming request. The SQL I slacked you creates the website if it doesn't already exist.

File: db/migrations/20201109152715_add_screenshots_table.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
+const { addUpdateTsTrigger } = require("../util")
+
+exports.up = async (knex) => {
+  await knex.schema.createTable("screenshots", (table) => {
```
Comment (will-weiss) [inline db/migrations/20201109152715_add_screenshots_table.js]: Let's call this `feedback_images` and, at somepoint, rename this in other parts of the system as well

File: db/migrations/20201109152715_add_screenshots_table.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    table.increments()
+    table.string("name").notNullable()
+    table
+      .integer("feedback_id")
+      .notNullable()
+      .references("id")
+      .inTable("feedbacks")
+      .onDelete("CASCADE")
+      .index()
+    table.binary("screenshot_file").notNullable()
```
Comment (will-weiss) [inline db/migrations/20201109152715_add_screenshots_table.js]: just `file` or `file_blob`

File: db/migrations/20201109152715_add_screenshots_table.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
+const { addUpdateTsTrigger } = require("../util")
+
+exports.up = async (knex) => {
+  await knex.schema.createTable("screenshots", (table) => {
+    table.increments()
+    table.string("name").notNullable()
```
Comment (will-weiss) [inline db/migrations/20201109152715_add_screenshots_table.js]: Should we include `file_extension`?

File: src/handlers.ts [hunk 6]
```diff
@@ -108,13 +112,36 @@ export const postFeedback = async (ctx: IRouterContext): Promise<any> => {
     throw { status: 401 }
   }
 
+  const [feedback] = await db<Feedback>('feedbacks').insert({ status, user_id: user.id }).returning('*')
```
Comment (will-weiss) [inline src/handlers.ts]: As discussed, let's grab the encodings for each of the files so we can do this in one transaction

File: db/migrations/20201109145014_add_feedbacks_table.js [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+    table.timestamps(true, true)
+  })
+
```
Comment (will-weiss) [inline db/migrations/20201109145014_add_feedbacks_table.js]: Let's include `tweetUrl`, a string

File: src/handlers.ts [hunk 8]
```diff
@@ -108,13 +112,36 @@ export const postFeedback = async (ctx: IRouterContext): Promise<any> => {
...
+
+  // Write all screenshots to the current directory
+  // const scrshts = await db<Screenshot>('screenshots').select('*')
+  // scrshts.forEach(s => {
+  //   writeFile(`./${s.name}.png`, s.screenshot_file, { encoding: 'base64' }, () => {
+  //     console.log('saved')
+  //   })
+  // })
+
+  const { url } = { url: 'fake' } // await twitter.tweetStatus(params)
```
Comment (will-weiss) [inline src/handlers.ts]: As far as handling twitter and/or our database potentially being down, how does this sound for wrangling our various data stores.

Try posting to twitter
  -> If that's successful, immediately reply to the user with the tweet URL and, in the background, save the feedback in the database as a single transaction with the returned tweetUrl. If that background process fails, log the error.
  -> If that's unsuccessful, log the error and immediately reply to the user saying the tweet was not successful. The client should then retry at a later point in time. Do not save anything to the database.

If we followed this approach, the `tweetUrl` would be a non-nullable unique index for the feedback table.

File: db/migrations/20201109145014_add_feedbacks_table.js (hunk 1 above)
Comment (shachar-langer) [reply db/migrations/20201109145014_add_feedbacks_table.js]: "feedback" is indeed plural, but we're talking here about rows/objects. I think that using "feedback" might confuse future maintainers.
(I also think that all the table names should be singular, but that's a different discussion)

File: db/migrations/20201109152715_add_screenshots_table.js (hunk 5 above)
Comment (shachar-langer) [reply db/migrations/20201109152715_add_screenshots_table.js]: At the moment we only save `png` files. I'm a fan of not adding unnecessary info to our databases. We can always use a migration script to add a `file_extension` column.

File: src/handlers.ts (hunk 6 above)
Comment (shachar-langer) [reply src/handlers.ts]: I'm not sure that inserting the encoding would work. Why not insert a binary?

File: src/handlers.ts [hunk 9]
```diff
@@ -40,12 +42,20 @@ const hostOf = (url: string): string => {
   }
 }
 
+function getCurrentUser(ctx: IRouterContext): SerializedUser {
```
Comment (will-weiss) [inline src/handlers.ts:45]: 👍

**Review by will-weiss — APPROVED** [review]

**Pull request merged**